File: customview/src/main/java/com/example/customview/adapter/HoverItemAdapter3.java
package com.example.customview.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.example.customview.R;
import com.example.customview.bean.HoverItemModel;

import org.jetbrains.annotations.NotNull;

import java.util.List;

public class HoverItemAdapter3 extends RecyclerView.Adapter<HoverItemAdapter3.MyViewHolder> {
    public static final int HAS_STICKY_VIEW = 1;
    public static final int NONE_STICKY_VIEW = 2;

    private final Context mContext;
    private final List<HoverItemModel> mList;

    public HoverItemAdapter3(Context mContext, List<HoverItemModel> list) {
        this.mContext = mContext;
        this.mList = list;
    }

    @NotNull
    @Override
    public MyViewHolder onCreateViewHolder(@NotNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(mContext).inflate(R.layout.layout_list_item, parent, false);
        return new MyViewHolder(view);
    }

    @Override
    public void onBindViewHolder(MyViewHolder viewHolder, int position) {
        HoverItemModel hoverItemModel = mList.get(position);
        viewHolder.tvName.setText(hoverItemModel.name);

        if (position == 0 || !hoverItemModel.sticky.equals(mList.get(position - 1).sticky)) {
            viewHolder.tvHeader.setVisibility(View.VISIBLE);
            viewHolder.tvHeader.setText(hoverItemModel.sticky);
            viewHolder.itemView.setTag(HAS_STICKY_VIEW);
        } else {
            viewHolder.tvHeader.setVisibility(View.GONE);
            viewHolder.itemView.setTag(NONE_STICKY_VIEW);
        }
        viewHolder.itemView.setContentDescription(hoverItemModel.sticky);
    }

    @Override
    public int getItemCount() {
        return mList == null ? 0 : mList.size();
    }

    public static class MyViewHolder extends RecyclerView.ViewHolder {
        public TextView tvHeader;
        public TextView tvName;

        public MyViewHolder(View itemView) {
            super(itemView);
            tvHeader = (TextView) itemView.findViewById(R.id.header_view);
            tvName = (TextView) itemView.findViewById(R.id.tv_name);
        }
    }
}

File: customview/src/main/java/com/example/customview/adapter/HoverItemAdapter3.kt
package com.example.customview.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.customview.bean.HoverItemModel
import com.example.customview.databinding.LayoutListItemBinding
import com.example.customview.utils.DisplayUtils

/**
 * 封装 Hover Item Adapter 3 相关逻辑的类。
 */

open class HoverItemAdapter3(
    private val mContext: Context?,
    private val mList: List<HoverItemModel>?
) : RecyclerView.Adapter<HoverItemAdapter3.MyViewHolder>() {

    companion object {
        const val HAS_STICKY_VIEW = 1
        const val NONE_STICKY_VIEW = 2
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {
        val binding = LayoutListItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return MyViewHolder(binding)
    }

    override fun onBindViewHolder(viewHolder: MyViewHolder, position: Int) {
        val hoverItemModel: HoverItemModel = mList!![position]
        viewHolder.tvName.text = hoverItemModel.name
        if ((position == 0) || hoverItemModel.sticky!! != mList[position - 1].sticky) {
            viewHolder.tvHeader.visibility = View.VISIBLE
            viewHolder.tvHeader.text = hoverItemModel.sticky
            viewHolder.itemView.tag = HAS_STICKY_VIEW
        } else {
            viewHolder.tvHeader.visibility = View.GONE
            viewHolder.itemView.tag = NONE_STICKY_VIEW
        }
        viewHolder.itemView.setContentDescription(hoverItemModel.sticky)
    }

    override fun getItemCount(): Int {
        return mList?.size ?: 0
    }

    open class MyViewHolder(binding: LayoutListItemBinding) : RecyclerView.ViewHolder(binding.root) {
        val tvHeader: TextView = binding.headerView
        val tvName: TextView = binding.tvName
    }
}
